ts/Data/Formula/FormulaTypes.ts

```diff
@@ -0,0 +1,287 @@
+/* *
+ *
+ *  (c) 2009-2023 Highsoft AS
+ *
+ *  License: www.highcharts.com/license
+ *
+ *  !!!!!!! SOURCE GETS TRANSPILED BY TYPESCRIPT. EDIT TS FILE ONLY. !!!!!!!
+ *
+ *  Authors:
+ *  - Sophie Bremer
+ *
+ * */
+
+
+'use strict';
+
+
+/* *
+ *
+ *  Declarations
+ *
+ * */
+
+
+/**
+ * Arguments array of a formula function with terms and ranges.
+ */
+export type Arguments = Array<(Range|Term)>;
+
+
+/**
+ * Formula array with terms and operators.
+ */
+export type Formula = Array<(Operator|Term)>;
+
+
+/**
+ * Formula function with an arguments array.
+ */
+export interface Function {
+    args: Arguments;
+    name: string;
+    type: 'function';
+}
+
+
+/**
+ * Item in arguments and formulas.
+ */
+export type Item = (Operator|Range|Term);
+
+
+/**
+ * Formula operators, either basic arithmetic, or basic logic.
+ */
+export type Operator = ('+'|'-'|'*'|'/'|'^'|'='|'<'|'<='|'>'|'>=');
+
+
+/**
+ * Represents a range to cells of a table.
+ */
+export interface Range {
+    beginColumn: number;
+    beginRow: number;
+    columnsRelative?: true;
+    endColumn: number;
+    endRow: number;
+    rowsRelative?: true;
+    type: 'range';
+}
+
+
+/**
+ * Represents a reference to a table cell.
+ */
+export interface Reference {
+    column: number;
+    columnRelative?: true;
+    row: number;
+    rowRelative?: true;
+    type: 'reference';
+}
+
+
+/**
+ * A term represents some form of processing into a value or is already a value.
+ */
+export type Term = (Formula|Function|Reference|Value);
+
+
+/**
+ * A value to use in a operation or process.
+ */
+export type Value = (boolean|number|string);
+
+
+/* *
+ *
+ *  Constants
+ *
+ * */
+
+
+/**
+ * Array of all possible operators.
+ * @private
+ */
+const operators: Array<string> =
+    ['+', '-', '*', '/', '^', '=', '<', '<=', '>', '>='];
+
+
+/* *
+ *
+ *  Functions
+ *
+ * */
+
+
+/**
+ * Converts non-number types to number.
+ *
+ * @param {Highcharts.FormulaValue} value
+ * Value to convert.
+ *
+ * @return {number}
+ * Number value. `NaN` if not convertable.
```

Comment:
From the description, it should convert, however, it converts strings to 0. Is this expected? And never returns NaN: (except when sending NaN)

Tested with:
```ts
> Highcharts._modules['Data/Formula/FormulaTypes.js'].asNumber('103')
> 0

// NaN tests:
> Highcharts._modules['Data/Formula/FormulaTypes.js'].asNumber(undefined);
> undefined

> Highcharts._modules['Data/Formula/FormulaTypes.js'].asNumber(NaN)
> NaN
```

Reply:
Edit:
Just realized it's a part of AVERAGEA implementation - the name is a bit misleading (especially we have another `asNumber` method in DataConverter). And this is used in `basicOperation`, which seems to be incorrect, this is g-spreadsheet:
<img width="406" alt="image" src="https://user-images.githubusercontent.com/1453926/236242293-4c5d3cbc-5958-4c2a-9cd1-f1403c44563e.png">
In our implementation we treat strings as 0's, so results will be: `1`, `FALSE` and `3`, right?